compiler/src/scala/quoted/runtime/impl/QuoteMatcher.scala

```diff
@@ -465,15 +505,31 @@ object QuoteMatcher {
           case Block(List(DefDef(nme.ANON_FUN, _, _, Apply(Ident(name), _))), _) => name.asTermName
           case arg => arg.symbol.name.asTermName
         }
-        val paramTypes = args.map(x => mapTypeHoles(x.tpe.widenTermRefExpr))
+        val paramTypes = args.map(x => adaptTypes(mapTypeHoles(x.tpe.widenTermRefExpr)))
         val methTpe = MethodType(names)(_ => paramTypes, _ => mapTypeHoles(patternTpe))
         val meth = newAnonFun(ctx.owner, methTpe)
         def bodyFn(lambdaArgss: List[List[Tree]]): Tree = {
           val argsMap = args.view.map(_.symbol).zip(lambdaArgss.head).toMap
           val body = new TreeMap {
             override def transform(tree: Tree)(using Context): Tree =
               tree match
-                case tree: Ident => env.get(tree.symbol).flatMap(argsMap.get).getOrElse(tree)
+                /*
+                 * When matching a method call `f(0)` against a HOAS pattern `p(g)` where
+                 * f has a method type `(x: Int): Int` and  `f` maps to `g`, `p` should hold
+                 * `g.apply(0)` because the type of `g` is `Int => Int` due to eta expansion.
+                 *
+                 * Remaining TODOs from issue-17105
+                 * * [x] cover the case of nested method call
+                 * * [ ] contextual params?
+                 * * [ ] erasure types?
+                 * * [x] eta-expand only HOAS param methods
+                 */
+                case Apply(_, _) if !tree.tpe.isInstanceOf[MethodicType] && env.contains(tree.symbol) =>
+                  def adaptMethodCalls(t: Tree): Tree = t match
```

Comment:
This code assumes that `tree` in this case is always in the form of `f(arg1)...(argn)`. Is it a safe assumption?

Reply:
Yes. If not all arguments are applied in the source, the application is eta-expanded to provide the missing arguments explicitly to the call.